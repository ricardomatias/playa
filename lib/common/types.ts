export enum BinaryEvent {
	Rest = 0,
	Hit = 1
}

export type Octaves = [number, number]

export enum RhythmType {
	Free = 'Free',
	Turn = 'Turn'
}

export type DistributionFunction = <T>(distribution: T | T[], precision?: number | undefined) => string[];
