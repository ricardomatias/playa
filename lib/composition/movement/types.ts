import { Interval, NoteSymbol, ScaleName, TurnMoves } from '../../constants';
import { GreekModeIntervals, ModePosition } from '../../constants/modes';

export const EUCLIDEAN = 'euclid';
export const TURN = 'turn';

export enum MovementRhythm {
	Euclidean = 'Euclidean',
	Turn = 'Turn'
}

export interface TimelineEventKey { root: NoteSymbol, scale: GreekModeIntervals }

export interface TimelineEvent {
	time: number;
	dur: number;
	key: TimelineEventKey;
}

export enum ModulationEventType {
	Mode = 'mode',
	Key = 'key'
}

export interface ModulationEvent {
	key: NoteSymbol;
	root: NoteSymbol;
	type: ModulationEventType;
	scaleName: ScaleName;
	position: ModePosition;
}

export type Movement = {
	timeline: TimelineEvent[];
	events: ModulationEvent[];
}

export interface Turn {
	type: TurnMoves,
	interval: typeof TurnMoves.Free | Interval | number | string
}

/**
 * @typedef {Object} Turn
 * @property {String} type '▼'
 * @property {String | Number} interval '⟷'
 * @example
 * { type: '▼', interval: '⟷' }
*/
export interface TurnEvent extends ModulationEvent {
	turn: Turn
}

export type TurnMovement = {
	timeline: TimelineEvent[];
	events: TurnEvent[];
}
