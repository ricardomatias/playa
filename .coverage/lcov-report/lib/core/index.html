<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/core</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/core
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">88.17% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>343/389</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">81.98% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>141/172</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">83.82% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>57/68</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">87.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>332/378</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="Chord.js"><a href="Chord.js.html">Chord.js</a></td>
	<td data-value="96.9" class="pic high"><div class="chart"><div class="cover-fill" style="width: 96%;"></div><div class="cover-empty" style="width:4%;"></div></div></td>
	<td data-value="96.9" class="pct high">96.9%</td>
	<td data-value="129" class="abs high">125/129</td>
	<td data-value="90.91" class="pct high">90.91%</td>
	<td data-value="44" class="abs high">40/44</td>
	<td data-value="94.44" class="pct high">94.44%</td>
	<td data-value="18" class="abs high">17/18</td>
	<td data-value="96.77" class="pct high">96.77%</td>
	<td data-value="124" class="abs high">120/124</td>
	</tr>

<tr>
	<td class="file high" data-value="HarmonyBase.js"><a href="HarmonyBase.js.html">HarmonyBase.js</a></td>
	<td data-value="84.62" class="pic high"><div class="chart"><div class="cover-fill" style="width: 84%;"></div><div class="cover-empty" style="width:16%;"></div></div></td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	<td data-value="66.67" class="pct medium">66.67%</td>
	<td data-value="6" class="abs medium">4/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="84.62" class="pct high">84.62%</td>
	<td data-value="13" class="abs high">11/13</td>
	</tr>

<tr>
	<td class="file high" data-value="Key.js"><a href="Key.js.html">Key.js</a></td>
	<td data-value="94.67" class="pic high"><div class="chart"><div class="cover-fill" style="width: 94%;"></div><div class="cover-empty" style="width:6%;"></div></div></td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	<td data-value="83.33" class="pct high">83.33%</td>
	<td data-value="36" class="abs high">30/36</td>
	<td data-value="92.31" class="pct high">92.31%</td>
	<td data-value="13" class="abs high">12/13</td>
	<td data-value="94.67" class="pct high">94.67%</td>
	<td data-value="75" class="abs high">71/75</td>
	</tr>

<tr>
	<td class="file low" data-value="Mode.js"><a href="Mode.js.html">Mode.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="9" class="abs low">0/9</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="24" class="abs low">0/24</td>
	</tr>

<tr>
	<td class="file high" data-value="Note.js"><a href="Note.js.html">Note.js</a></td>
	<td data-value="89.74" class="pic high"><div class="chart"><div class="cover-fill" style="width: 89%;"></div><div class="cover-empty" style="width:11%;"></div></div></td>
	<td data-value="89.74" class="pct high">89.74%</td>
	<td data-value="78" class="abs high">70/78</td>
	<td data-value="90.48" class="pct high">90.48%</td>
	<td data-value="42" class="abs high">38/42</td>
	<td data-value="84.21" class="pct high">84.21%</td>
	<td data-value="19" class="abs high">16/19</td>
	<td data-value="89.61" class="pct high">89.61%</td>
	<td data-value="77" class="abs high">69/77</td>
	</tr>

<tr>
	<td class="file high" data-value="Scale.js"><a href="Scale.js.html">Scale.js</a></td>
	<td data-value="93.94" class="pic high"><div class="chart"><div class="cover-fill" style="width: 93%;"></div><div class="cover-empty" style="width:7%;"></div></div></td>
	<td data-value="93.94" class="pct high">93.94%</td>
	<td data-value="66" class="abs high">62/66</td>
	<td data-value="82.86" class="pct high">82.86%</td>
	<td data-value="35" class="abs high">29/35</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="10" class="abs high">8/10</td>
	<td data-value="93.44" class="pct high">93.44%</td>
	<td data-value="61" class="abs high">57/61</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file high" data-value="types.js"><a href="types.js.html">types.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
