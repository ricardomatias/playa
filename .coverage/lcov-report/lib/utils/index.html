<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for lib/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> lib/utils
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">90.6% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>106/117</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">91.11% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>41/45</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">79.17% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>19/24</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">90.83% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>99/109</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line high'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file high" data-value="chord.js"><a href="chord.js.html">chord.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="6" class="abs high">6/6</td>
	</tr>

<tr>
	<td class="file high" data-value="combinations.js"><a href="combinations.js.html">combinations.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="18" class="abs high">18/18</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="16" class="abs high">16/16</td>
	</tr>

<tr>
	<td class="file medium" data-value="functional.js"><a href="functional.js.html">functional.js</a></td>
	<td data-value="76.47" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 76%;"></div><div class="cover-empty" style="width:24%;"></div></div></td>
	<td data-value="76.47" class="pct medium">76.47%</td>
	<td data-value="17" class="abs medium">13/17</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="55.56" class="pct medium">55.56%</td>
	<td data-value="9" class="abs medium">5/9</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="12" class="abs medium">9/12</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

<tr>
	<td class="file medium" data-value="map.js"><a href="map.js.html">map.js</a></td>
	<td data-value="70" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 70%;"></div><div class="cover-empty" style="width:30%;"></div></div></td>
	<td data-value="70" class="pct medium">70%</td>
	<td data-value="20" class="abs medium">14/20</td>
	<td data-value="57.14" class="pct medium">57.14%</td>
	<td data-value="7" class="abs medium">4/7</td>
	<td data-value="75" class="pct medium">75%</td>
	<td data-value="4" class="abs medium">3/4</td>
	<td data-value="68.42" class="pct medium">68.42%</td>
	<td data-value="19" class="abs medium">13/19</td>
	</tr>

<tr>
	<td class="file high" data-value="note.js"><a href="note.js.html">note.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="4" class="abs high">4/4</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="2" class="abs high">2/2</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

<tr>
	<td class="file high" data-value="octaves.js"><a href="octaves.js.html">octaves.js</a></td>
	<td data-value="97.5" class="pic high"><div class="chart"><div class="cover-fill" style="width: 97%;"></div><div class="cover-empty" style="width:3%;"></div></div></td>
	<td data-value="97.5" class="pct high">97.5%</td>
	<td data-value="40" class="abs high">39/40</td>
	<td data-value="95.24" class="pct high">95.24%</td>
	<td data-value="21" class="abs high">20/21</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="97.5" class="pct high">97.5%</td>
	<td data-value="40" class="abs high">39/40</td>
	</tr>

<tr>
	<td class="file high" data-value="whilst.js"><a href="whilst.js.html">whilst.js</a></td>
	<td data-value="100" class="pic high"><div class="chart"><div class="cover-fill cover-full" style="width: 100%;"></div><div class="cover-empty" style="width:0%;"></div></div></td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="3" class="abs high">3/3</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="1" class="abs high">1/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="8" class="abs high">8/8</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Wed Feb 20 2019 15:09:45 GMT+0100 (Central European Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
