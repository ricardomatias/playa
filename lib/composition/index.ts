export { createMotif } from './motif';
export { createChordProgression } from './chord-progression';
export { createMelodies } from './melodies';
export { createPercussion } from './percussion';
export { createArp, genHarmonicShift } from './arp';

export * as Rhythm from './rhythm';
export * from './sequence';
export * from './movement';
