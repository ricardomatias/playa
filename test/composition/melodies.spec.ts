import * as R from 'ramda';
import { Key } from '../../lib/core';
import { createMovement, createMelodies } from '../../lib/composition';
import * as Rhythm from '../../lib/composition/rhythm';
import random from '../../lib/tools/random';
import { distribute } from '@ricardomatias/roll';
import { RhythmType } from '../../lib/common/types';
import { valuesToArr } from '../../lib/utils';

describe('A Melodies test suite', () => {
	it('should generate melodies based on a movement timeline - free rhythm', () => {
		// given
		random.setSeed('test');

		const aMaj = new Key('A', Key.Major);

		const opts = {
			modProb: 0.4,
		};

		// when
		const movement = createMovement(aMaj, '4:0:0', 6, opts);

		const melodies = createMelodies(movement.timeline);

		expect(melodies).toMatchSnapshot();
	});

	it('should generate melodies based on a movement timeline - turn rhythm', () => {
		// given
		random.setSeed('test');

		const aMaj = new Key('A', Key.Major);

		const opts = {
			modProb: 0.4,
		};

		// when
		const movement = createMovement(aMaj, '4:0:0', 8, opts);

		const melodies = createMelodies(movement.timeline, {
			rhythmType: RhythmType.Turn,
		});

		expect(melodies).toMatchSnapshot();
	});

	it('should generate progression with defined structures', () => {
		// given
		random.setSeed('test');

		const aMaj = new Key('Db', Key.Major);

		const opts = {
			modProb: 0.4,
		};

		// when
		const movement = createMovement(aMaj, '2:0:0', 10, opts);

		const melodies = createMelodies(movement.timeline, {
			rhythmValues: R.flatten(valuesToArr(Rhythm.Presets)),
			distribution: distribute.decreasing,
			restProb: 0.5,
			octaves: [
				[2, 1],
				[3, 1],
			],
		});

		expect(melodies).toMatchSnapshot();
	});
});
