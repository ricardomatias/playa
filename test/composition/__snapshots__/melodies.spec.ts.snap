// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A Melodies test suite should generate melodies based on a movement timeline - free rhythm 1`] = `
Array [
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 73,
    "next": 240,
    "note": "C#4",
    "time": 0,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 71,
    "next": 960,
    "note": "B3",
    "time": 240,
  },
  Object {
    "dur": 480,
    "isRest": false,
    "midi": 73,
    "next": 1440,
    "note": "C#4",
    "time": 960,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 64,
    "next": 2160,
    "note": "E3",
    "time": 1440,
  },
  Object {
    "dur": 480,
    "isRest": false,
    "midi": 64,
    "next": 2640,
    "note": "E3",
    "time": 2160,
  },
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 61,
    "next": 2880,
    "note": "C#3",
    "time": 2640,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 69,
    "next": 3600,
    "note": "A3",
    "time": 2880,
  },
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 71,
    "next": 3840,
    "note": "B3",
    "time": 3600,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 78,
    "next": 4560,
    "note": "Gb4",
    "time": 3840,
  },
  Object {
    "dur": 480,
    "isRest": false,
    "midi": 78,
    "next": 5040,
    "note": "Gb4",
    "time": 4560,
  },
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 75,
    "next": 5280,
    "note": "Eb4",
    "time": 5040,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 64,
    "next": 6000,
    "note": "E3",
    "time": 5280,
  },
  Object {
    "dur": 480,
    "isRest": false,
    "midi": 64,
    "next": 6480,
    "note": "E3",
    "time": 6000,
  },
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 61,
    "next": 6720,
    "note": "Db3",
    "time": 6480,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 63,
    "next": 7440,
    "note": "Eb3",
    "time": 6720,
  },
  Object {
    "dur": 240,
    "isRest": false,
    "midi": 64,
    "next": 7680,
    "note": "E3",
    "time": 7440,
  },
]
`;

exports[`A Melodies test suite should generate melodies based on a movement timeline - turn rhythm 1`] = `
Array [
  Object {
    "dur": 320,
    "isRest": false,
    "midi": 73,
    "next": 320,
    "note": "C#4",
    "time": 0,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 71,
    "next": 480,
    "note": "B3",
    "time": 320,
  },
  Object {
    "dur": 480,
    "isRest": false,
    "midi": 73,
    "next": 960,
    "note": "C#4",
    "time": 480,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 64,
    "next": 1120,
    "note": "E3",
    "time": 960,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 64,
    "next": 1200,
    "note": "E3",
    "time": 1120,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 61,
    "next": 1920,
    "note": "C#3",
    "time": 1200,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 71,
    "next": 2080,
    "note": "B3",
    "time": 1920,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 71,
    "next": 2160,
    "note": "B3",
    "time": 2080,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 68,
    "next": 2880,
    "note": "Ab3",
    "time": 2160,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 66,
    "next": 3040,
    "note": "Gb3",
    "time": 2880,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 66,
    "next": 3120,
    "note": "Gb3",
    "time": 3040,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 63,
    "next": 3840,
    "note": "Eb3",
    "time": 3120,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 64,
    "next": 4000,
    "note": "E3",
    "time": 3840,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 64,
    "next": 4080,
    "note": "E3",
    "time": 4000,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 61,
    "next": 4800,
    "note": "Db3",
    "time": 4080,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 64,
    "next": 4960,
    "note": "E3",
    "time": 4800,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 64,
    "next": 5040,
    "note": "E3",
    "time": 4960,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 61,
    "next": 5760,
    "note": "Db3",
    "time": 5040,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 68,
    "next": 5920,
    "note": "G#3",
    "time": 5760,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 68,
    "next": 6000,
    "note": "G#3",
    "time": 5920,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 64,
    "next": 6720,
    "note": "E3",
    "time": 6000,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 66,
    "next": 6880,
    "note": "Gb3",
    "time": 6720,
  },
  Object {
    "dur": 80,
    "isRest": false,
    "midi": 66,
    "next": 6960,
    "note": "Gb3",
    "time": 6880,
  },
  Object {
    "dur": 720,
    "isRest": false,
    "midi": 63,
    "next": 7680,
    "note": "Eb3",
    "time": 6960,
  },
]
`;

exports[`A Melodies test suite should generate progression with defined structures 1`] = `
Array [
  Object {
    "dur": 320,
    "isRest": true,
    "midi": -1,
    "next": 320,
    "note": "",
    "time": 0,
  },
  Object {
    "dur": 160,
    "isRest": false,
    "midi": 61,
    "next": 480,
    "note": "Db3",
    "time": 320,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 60,
    "next": 840,
    "note": "C3",
    "time": 480,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 960,
    "note": "",
    "time": 840,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 67,
    "next": 1320,
    "note": "G3",
    "time": 960,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 1440,
    "note": "",
    "time": 1320,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 74,
    "next": 1800,
    "note": "D4",
    "time": 1440,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 1920,
    "note": "",
    "time": 1800,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 72,
    "next": 2280,
    "note": "C4",
    "time": 1920,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 2400,
    "note": "",
    "time": 2280,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 72,
    "next": 2760,
    "note": "C4",
    "time": 2400,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 2880,
    "note": "",
    "time": 2760,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 75,
    "next": 3240,
    "note": "Eb4",
    "time": 2880,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 3360,
    "note": "",
    "time": 3240,
  },
  Object {
    "dur": 360,
    "isRest": false,
    "midi": 74,
    "next": 3720,
    "note": "D4",
    "time": 3360,
  },
  Object {
    "dur": 120,
    "isRest": true,
    "midi": -1,
    "next": 3840,
    "note": "",
    "time": 3720,
  },
]
`;
